packages/gql-executor/gqlExecutor.ts

```diff
@@ -1,27 +1,58 @@
-import "./tracer"
-import '../server/initSentry'
 import Redis from 'ioredis'
 import {ServerChannel} from 'parabol-client/types/constEnums'
+import GQLExecutorId from '../client/shared/gqlIds/GQLExecutorId'
+import executeGraphQL, {GQLRequest} from '../server/graphql/executeGraphQL'
+import '../server/initSentry'
+import RedisStream from './RedisStream'
+import './tracer'
 
-const {REDIS_URL} = process.env
-const publisher = new Redis(REDIS_URL)
-const subscriber = new Redis(REDIS_URL)
-
+const {REDIS_URL, SERVER_ID} = process.env
 interface PubSubPromiseMessage {
   jobId: string
-  request: Record<string, unknown>
+  request: GQLRequest
 }
 
-const onMessage = async (_channel: string, message: string) => {
-  const {jobId, request} = JSON.parse(message) as PubSubPromiseMessage
-  const executeGraphQL = require('../server/graphql/executeGraphQL').default
-  const response = await executeGraphQL(request)
-  publisher.publish(
-    ServerChannel.GQL_EXECUTOR_RESPONSE,
-    JSON.stringify({response, jobId})
+const run = async () => {
+  const subscriber = new Redis(REDIS_URL)
+  const publisher = new Redis(REDIS_URL)
+  const serverChannel = GQLExecutorId.join(SERVER_ID)
+
+  // subscribe to direct messages
+  const onMessage = async (_channel: string, message: string) => {
+    const {jobId, request} = JSON.parse(message) as PubSubPromiseMessage
+    const response = await executeGraphQL(request)
+    publisher.publish(ServerChannel.GQL_EXECUTOR_RESPONSE, JSON.stringify({response, jobId}))
+  }
+
+  subscriber.on('message', onMessage)
+  subscriber.subscribe(serverChannel)
+
+  // subscribe to consumer group
+  try {
+    await publisher.xgroup(
+      'CREATE',
+      ServerChannel.GQL_EXECUTOR_STREAM,
+      ServerChannel.GQL_EXECUTOR_CONSUMER_GROUP,
+      '$',
+      'MKSTREAM'
+    )
+  } catch (e) {
+    // stream already exists
+  }
+
+  const incomingStream = new RedisStream(
+    ServerChannel.GQL_EXECUTOR_STREAM,
+    ServerChannel.GQL_EXECUTOR_CONSUMER_GROUP,
+    serverChannel
   )
+  console.log(`\n💧💧💧 Ready for GraphQL Execution: ${SERVER_ID} 💧💧💧`)
+
+  for await (const message of incomingStream) {
+    // don't await the call below so this instance can immediately call incomingStream.next()
+    // and be put back in the consumer group, which means it can process more than 1 job at a time
+    // See https://www.loom.com/share/b56812bc561348d0b3d74fe35414499d
+    onMessage('', message)
```

Comment:
I wanted to disagree here, but did some testing and it seems I was wrong and it's fine :+1: 
https://www.loom.com/share/0d8aeea802674a76b0c658e3bcdce7c8

Reply:
thank you for the sanity check!
i was thinking about keeping a counter of how many jobs are currently processing, but i figured monitoring CPU load will do the same thing without setting an arbitrarily limit